### ISSUE TITLE
[BatchNorm] Add boolean flags to choose the stats for normalization

### ISSUE THREAD

**Issue by frgfm** [pytorch/pytorch#39674]
## 🚀 Flags for flexibility in BN
Hello there :wave: 

Following up on discussions with @SsnL in #37823 and #38084, I believe some clarifications are indeed in the behaviour of the `forward` method of Batch normalization layers.

Concisely, the point of this issue would be to discuss the following points:
1. ability to dynamically tell the layer whether it should update its population stats' estimators (buffer) or not
2. ability to dynamically tell the layer whether the stats used for normalization should be the actual current mini-batch stats or the already seen population's estimators (buffer).
3. the train --> eval switch should use the computed population mean & variance for normalization if they are available, prevent population estimators from updating, and avoid updating the affine parameters.

## Motivation

Initially, I was only looking for the best way to use `BatchNorm2d` in frozen layers (in the case of transfer learning). Obviously, `FrozenBatchNorm2d` from torchvision was a good option but required replacing all BNs in a model (and there was no eps in frozenBN until very recently). I went looking for a simple flag that would work in vain (this needs to apply for both train and eval modes, otherwise the frequent switches that occur during a training procedure make it irrelevant)

## Pitch

First, `track_running_stats` should be converted into a constructor argument but not an attribute since it can be switched by the user later (meaning we cannot dynamically rely on this to assume whether the buffer is available and/or has been tracked)

Second, we could introduce two dynamic flags:
- `update_running_stats` (only renaming the original track_running_stats actually): controls whether the buffer should be updated on forward
- `minibatch_normalization`: controls whether the mini-batch stats should be used for data normalization instead of population estimators (buffer) during forward.

Since, in the current API, switching to train/eval mode does not change the parameters' requires_grad then:

- updating population stats (buffer) is only done if `self.update_running_stats and self.training`
- normalization is done with mini-batch stats rather than buffer's if `self.minibatch_normalization or self.training`

**Comment by frgfm** [comment]
Any thoughts @albanD @mruberry?

**Comment by albanD** [comment]
Hi,

Sorry for the delay.
I definitely agree that we want to clean this up!

Two questions I would have:
- Do I understand properly that when self.training=False, we can never update the running stats and have to use the saved statistics for normalization?
- Is it BC? For a user that just uses batchnorm today, will there be any change?

**Comment by mruberry** [comment]
I'm also concerned about the BC implications of (3). (1) and (2) seem OK but I also know that changing batchnorm has historically been challenging in some cases, especially ensuring it's properly jittable. 

I'm not super familiar with the context here, but how much would implementing a torch.nn.batchnorm.functional address these issues? That would allow users the finest control over what statistics to use and what to update.

**Comment by frgfm** [comment]
@albanD 
- it's up for discussions but for I do indeed suggest that we enforce this behaviour when `self.training=False`
- with the default values I suggested, it's not a breaking change. It just allows for additional specific behaviour for advanced users.

@mruberry for the (3) I was talking about general training behaviour vs evaluation behaviour, not `self.training`. But I fully agree that this must not be a breaking change. Regarding its "jittability", I am happy conducting some test (with guidance) to make sure it works.

I agree that adding a functional API would help, but the points I raised are not really dependent on the interface in my point of view. For further context, I would suggest to quickly check my discussions with @SsnL in #37823 and #38084

Let me know if I can help somehow!

**Comment by mruberry** [comment]
I've finally had a chance to review this more thoroughly and I appreciate your proposal. The current behavior is a little convoluted.

To recap today: 

- Today, if "track running stats" is True then population statistics are tracked. This is the default. 
- If the batchnorm is in training mode and track running stats if False then population stats are also used (if available).
- If the batchnorm is in training mode then population stats are updated. Otherwise they are not.
- In training mode today the minibatch stats are used for normalization.
- In eval mode today the population stats are used for normalization, if they're available.

In your (3), however, you mention that putting batchnorm in eval mode should do three things: use the population stats, stop updating the population stats, and stop updating the affine parameters. Is it already doing some of these things in eval mode? 

I think the first part of your proposal will make which stats are used and when they are updated much clearer. And, as mentioned above, I think we can implement them without BC concerns. Would you like to submit a PR for them and separate changes to batchnorm's eval behavior for a follow-up? I think you could keep the "track running stats" flag as it is so only people who manipulate these new flags will see the change in behavior.

**Comment by ppwwyyxx** [comment]
IMO pytorch should include standard versions of standard layers. While there could be many different variants of BatchNorm, just like there could be variants of convolution, pooling, it wasn't so reasonable for pytorch to include them unless any of them become more standard.
The `num_batches_tracked` and `track_running_stats` features are examples of things that are non-standard but added in pytorch. As a result, they cause confusions about what they do & what they should do (like those linked by in this issue), lead to unnecessary complexity in other places (e.g. jittable), and lead to unnecessary performance penalty (https://github.com/pytorch/pytorch/issues/27926).

I do feel different variants of BatchNorm are useful and I've used >10 variants of it in my research, but I don't think pytorch is the right place to play with more of them.

**Comment by mruberry** [comment]
> IMO pytorch should include standard versions of standard layers. While there could be many different variants of BatchNorm, just like there could be variants of convolution, pooling, it wasn't so reasonable for pytorch to include them unless any of them become more standard.
> The `num_batches_tracked` and `track_running_stats` features are examples of things that are non-standard but added in pytorch. As a result, they cause confusions about what they do & what they should do (like those linked by in this issue), lead to unnecessary complexity in other places (e.g. jittable), and lead to unnecessary performance penalty (#27926).
> 
> I do feel different variants of BatchNorm are useful and I've used >10 variants of it in my research, but I don't think pytorch is the right place to play with more of them.

I don't think the intent of @frgfm's proposal is to make batchnorm less standard, it's simply to give users more control over how it operates.

**Comment by frgfm** [comment]
Thanks for diving into this @mruberry!

Regarding (3), to the best of my understanding, it is already doing this, I was just listing everything expected from this switch. Yes sure, I'll work on a PR shortly!

@ppwwyyxx As mruberry said, I do not aim at implementing a different version of BatchNorm, but rather focusing on two things:
- clarity of interface (knowing with what it will be normalized, and whether the population stats should be updated)
- flexibility of understudied usage of BN (namely, the issue that started all this was about what the BN behaviour should be in frozen layers and how to achieve this behaviour using the current arguments and attributes of the class)
Happy to discuss alternatives anyway!

**Comment by ppwwyyxx** [comment]
More clarity of the current interface would be a great improvement. I also feel that's needed on the current interface.
But "understudied usage of BN" is, in other words, non-standard usage of BN. There are many such understudied usage of layers: if you look at the [equivalent of BN in tensorflow](https://www.tensorflow.org/api_docs/python/tf/keras/layers/BatchNormalization), it also has such extra "features" of its own: renorm, virtual_batch_size, adjustment of mean/var. As you can imagine, everyone thinks the features they need are more suitable than others to be included. Among all possible variants, if there are rules to determine what pytorch should include and what not to, I think how standard & widely used something is should be the most important criteria.

Aside from whether pytorch should include these usages, a suggestion is that when I need variants of BN I often implement them as separate modules, just to avoid the huge combination space of [all the flags] x [train/val mode] x [requires_grad], etc. Previous issues from you such as https://github.com/pytorch/pytorch/issues/37823 and https://github.com/pytorch/pytorch/pull/38084 show how difficult it may become to maintain the combination space.

**Comment by mruberry** [comment]
> More clarity of the current interface would be a great improvement. I also feel that's needed on the current interface.
> But "understudied usage of BN" is, in other words, non-standard usage of BN. There are many such understudied usage of layers: if you look at the [equivalent of BN in tensorflow](https://www.tensorflow.org/api_docs/python/tf/keras/layers/BatchNormalization), it also has such extra "features" of its own: renorm, virtual_batch_size, adjustment of mean/var. As you can imagine, everyone thinks the features they need are more suitable than others to be included. Among all possible variants, if there are rules to determine what pytorch should include and what not to, I think how standard & widely used something is should be the most important criteria.
> 
> Aside from whether pytorch should include these usages, a suggestion is that when I need variants of BN I often implement them as separate modules, just to avoid the huge combination space of [all the flags] x [train/val mode] x [requires_grad], etc. Previous issues from you such as #37823 and #38084 show how difficult it may become to maintain the combination space.

That makes sense @ppwwyyxx, but these flags affect the current features. They're not really adding new Batchnorm features.

**Comment by ppwwyyxx** [comment]
Well, I guess adding any new flag counts as a new "feature" to me :smile:  but I see your point that the changes proposed here are more about controlling the existing behaviors rather than adding new behaviors (as opposed to, e.g., those flags in tensorflow).

**Comment by imjiangjun** [comment]
I think it's a very good proposal.
I have been confused by the flags in BN layer for a long time, especially in the situation mentioned in #37823. It's really strange that the `running_mean` and `running_var` still change even `track_running_stats` has been set to False.
As discussed above, there are three kinds of behavior to control:

1. Whether the **affine parameters** (`weight` and `bias`) should be **updated**.
2. Which **stat to use for normalization**: the actual current mini-batch stats or the already seen population's estimators (buffer, i.e., `running_mean` and `running_var`).
3. Whether the **buffer** (`running_mean` and `running_var`) should be **updated**.

Obviously, only **two** flags (`training` and `track_running_stats`) cannot control **three** kinds of behavior at the same time. I think it's the essential reason for what caused the confused situation such as #37823. **It can't be clarified until another flag is added.**
There is no need to worry about new features being introduced because nothing is introduced. **It just makes controlling the three existing features more clear.**

**Comment by frgfm** [comment]
Hey everyone :wave: 

Thanks a lot for all your thoughts on this topic!
I'm happy to work on a PR if we are all on the same page regarding the approach to take here. Here is my suggestion:

1. make `track_running_stats` a constructor argument only specifying whether running stats attributes should be initialized
2. create a dynamic flag `update_running_stats`: controls whether the buffer should be updated on forward in training mode
3. create a `minibatch_normalization` flag: controls whether the mini-batch stats should be used for data normalization instead of population estimators (buffer) during forward in eval mode (enforced in training mode).

If the above is OK with everyone, perhaps 2 PRs might be a good option:
- first PR to convert `track_running_stats` into a constructor argument and implement the behaviour of `update_running_stats`
- second PR to implement the behaviour of `minibatch_normalization`

Any thoughts?

**Comment by imjiangjun** [comment]
Maybe we should make changes carefully. There are many things to be considered. E.g., how many and what arguments should be added? Should the existing arguments' behavior be changed? What's the best default behavior? Should it be enforced in training mode or allowed to be changed? BN is so important that it needs more discussion.

**Comment by mruberry** [comment]
Sorry for the delay, @frgfm, it's been a busy time and this issue is a little complicated. We'll review it carefully soon.

**Comment by mruberry** [comment]
> 1. make `track_running_stats` a constructor argument only specifying whether running stats attributes should be initialized

This would actually be just a documentation change, right? `track_running_stats` is already a constructor argument. See. https://pytorch.org/docs/master/generated/torch.nn.BatchNorm1d.html?highlight=track_running_stats.

> 2. create a dynamic flag `update_running_stats`: controls whether the buffer should be updated on forward in training mode
> 3. create a `minibatch_normalization` flag: controls whether the mini-batch stats should be used for data normalization instead of population estimators (buffer) during forward in eval mode (enforced in training mode).

Adding functions for controlling these flags seems OK as long as we don't regress performance by checking them. We will also have to be careful about documenting what tracing/scripting limitations they have, if any. It would be OK to modify only the Python module in a first PR as long as it's possible to update the C++ module for consistency later. The default values for these flags should preserve the current behavior.

**Comment by mruberry** [comment]
> Maybe we should make changes carefully. 

We should!

E.g., how many and what arguments should be added? Should the existing arguments' behavior be changed? What's the best default behavior? Should it be enforced in training mode or allowed to be changed? 

Probably no new arguments and no change to existing or default behavior to avoid backwards-compatibility issues.

**Comment by frgfm** [comment]
@mruberry 
I suggest `track_running_stats` to be kept as constructor argument but not as a class attribute (if True, it would set the value of the attribute `update_running_stats` accordingly, which will be an attribute but not a constructor argument as per your last comment)

Should I make a first PR about this? (i.e., deprecating the ̀track_running_stats` attribute, only keeping it as a constructor arg + setting the `update_running_stats` attribute)

Then I could come up with a proposition for `minibatch_normalization` in a follow-up PR.

**Comment by imjiangjun** [comment]
@frgfm  Looking forward to your PR. And I have a suggestion.
> create a minibatch_normalization flag: controls whether the mini-batch stats should be used for data normalization instead of population estimators (buffer) during forward in eval mode (enforced in training mode).
I suggest that the `minibatch_normalization ` behavior in training mode described above is set **default** rather than **enforced**. When some layers including BN from other pretrained base model are frozen in traning mode, it's allowed to choose which stats to use for those BN layers in some special situations.

**Comment by frgfm** [comment]
@mruberry just checking whether you agree with the general idea of the potential PR?

@imjiangjun perhaps I misunderstood your suggestion, but that's one of the goal of my suggestion actually :ok_hand:

**Comment by mruberry** [comment]
@frgfm Looking through the prior comments I believe I did. It's been a long time, though. We even have new torch.nn engineers now! I'll mention this to them.

**Comment by frgfm** [comment]
> @frgfm Looking through the prior comments I believe I did. It's been a long time, though. We even have new torch.nn engineers now! I'll mention this to them.

Oh my bad, I wasn't sure you were OK with me opening a PR about this! Glad to hear more people are working on `torch.nn`, does this mean opening a PR wouldn't be worth it anymore?
Feel free to tag me in the upcoming discussions :ok_hand:
